add release notes for query iter sorts (#1460)
bevyengine/bevy-website: release-content/0.14/release-notes/13417_implement_the_full_set_of_sort_methods_on_QueryIter.md

> Currently, query iteration order is not guaranteed. If we wish to work with our query items in a certain order, we need sorting.

Reviewer: ```suggestion
Bevy does not make any guarantees about the order of items. So if we wish to work with our query items in a certain order, we need to sort them!
```

Flow / voice edit.

Author: Done.

> We might want sort our queries to f.e. display a list of things, or fold over our query with some order-dependent math operation.

Reviewer: ```suggestion
We might want to display the scores of the players in order, or ensure a consistent iteration order for the sake of networking stability.
```

f.e. isn't a particularly common abbreviation in English, and I've made the examples more concrete / game-relevant :)

Author: Done!

In 0.13 a sort could look like this:

```rust
#[derive(Component, Copy, Clone, Deref)]
pub struct Attack(pub usize)

fn handle_enemies(enemies: Query<(&Health, &Attack, &Defense)>) {

>     let mut enemies: Vec<_> = enemies.iter().collect();

Reviewer: ```suggestion
    // Oh no! An allocation!
    let mut enemies: Vec<_> = enemies.iter().collect();
```

Author: Added! Without the oh no, since you can't avoid allocation with sorts :)

    enemies.sort_by_key(|(_, atk, ..)| *atk)
    for enemy in enemies {
        work_with(enemy)
    }
}
```

This can get especially unwieldy and repetitive when sorting within multiple systems.
Even if we always want the same sort, different [`Query`] types make it unreasonably difficult to abstract away as a user!
To solve this, we implemented new sort methods on the [`QueryIter`] type, turning the example into:

```rust
// To be used as a sort key, `Attack` now implements Ord.
#[derive(Component, Copy, Clone, Deref, PartialEq, Eq, PartialOrd, Ord)]
pub struct Attack(pub usize)

fn handle_enemies(enemies: Query<(&Health, &Attack, &Defense)>) {

>     for enemy in enemies.iter().sort::<&Attack>() {

Reviewer: Does this also allocate under the hood? If so, we should call it out.

Author: A call-out has been added, (without mentioning the Vec type)

        work_with(enemy)
    }
}
```

To sort our query with the `Attack` component, we specify it as the generic parameter to [`sort`].

> If we wished to sort with more than one [`Component`], we can do so, independent of [`Component`] order in the original [`Query`] type: `enemies.iter().sort::<(&Defense, &Attack)>()`

Reviewer: ```suggestion
To sort by more than one [`Component` `enemies.iter().sort::<(&Defense, &Attack)>()`: first by `Defense`, then by `Attack`.
```

Clarity and brevity.

Author: Changed.

The generic parameter can be thought of as being a [lens] or "subset" of the original query, on which the underlying sort is actually performed. The result is then internally used to return a new sorted query iterator over the original query items.
With the default [`sort`], the lens has to be fully [`Ord`], like with [`slice::sort`].
If this is not enough, we also have the counterparts to the remaining 6 sort methods from [`slice`]!

The generic lens argument works the same way as with [`Query::transmute_lens`]. We do not use filters, they are inherited from the original query.
The [`transmute_lens`] infrastructure has some nice additional features, which allows for this:

```rust
fn handle_enemies(enemies: Query<(&Health, &Attack, &Defense, &Rarity)>) {
    for enemy in enemies.iter().sort_unstable::<Entity>() {
        work_with(enemy)
    }
}
```

Because we can add [`Entity`] to any lens, we can sort by it without including it in the original query!

> These sort methods are fully generic over mutability, so can be used on both [`Query::iter`] and [`Query::iter_mut`]! The rest of the of the iterator methods on [`Query`] do not support sorting.

Reviewer: ```suggestion
These sort methods work with both [`Query::iter`] and [`Query::iter_mut`]! The rest of the of the iterator methods on [`Query`] do not currently support sorting.
```
Brevity :)

Author: It might be worth aligning this with the fact that sorting of this kind can only ever be added to `iter_many` and `iter_many_mut` and none of the others.

Reviewer: I agree with you, but I think that's not relevant to the general audience reading this.

Author: Applied!

The sorts return [`QuerySortedIter`], itself an iterator, enabling the use of further iterator adapters on it.

> Further, [`QuerySortedIter`] implements [`DoubleEndedIterator`] while the initial [`QueryIter`] does not. As a consequence, an empty sort can be used to get a more powerful iterator type:

Reviewer: IMO all of the text between "Further..." and "Keep in mind" should be cut. These are really neat uses, but this is too much detail for the release notes. If it's not in there already, these should be in the docs though!

Author: Fair!

Author: The section has been removed.

// `StatisticsData` does not implement `Copy`.
#[derive(Component, Clone, Deref, PartialEq, Eq, PartialOrd, Ord)]
pub struct Statistics(pub Arc<StatisticData>)

// Does not compile.
fn show_stats(users: Query<(&User, &Statistics)>) {
   let mut users: Vec<_> = users.iter().collect();
   users.sort_by_key(|(_, stats)| *stats)
   show(users)
}

// Compiles.
fn show_stats_2(users: Query<(&User, &Statistics)>) {
   let users = users.iter().sort::<&Statistics>();
   show(users)
}
```

In current Rust, we can not return references from the key extraction closure in [`slice::sort_by_key`]/[`slice::sort_by_cached_key`].
This can become a headache when using non-[`Copy`] [`Component`]s.
The new [`sort`]/[`sort_by`] methods can work around this by having the lensing be the "key extraction", removing the need for such a closure.

Keep in mind that the lensing does add some overhead, so these query iterator sorts do not perform equally to a manual sort on average. However, this *strongly* depends on workload, so best test it yourself if relevant!
The sorts themselves are not yet cached between system runs, and these methods only sort iterators, not underlying query storage.

> Note that query iteration might happen in a deterministic order for now, but that may change anytime over future releases.

Reviewer: I think this should be cut.

Author: I cut the last 2 sentences. Some people care a lot about sort performance, so I think the mention of it should remain.
It is not some simple "sort sugar" like itertools, but that might not be clear.

> Currently, query iteration order is not guaranteed. If a user wishes to handle their query items in a certain order, sorting is required.

Reviewer: It would be nice to give an example or two about why in-order iteration is useful.

Author: added 2 examples!
